### PR TITLE
asymmetric laplace distribution added

### PR DESCRIPTION
**Pull request by chandan5362**
Asymmetric Laplace distribution added for  #4351
please do let me know if there are any unnecessary changes or any changes that are yet to be made.

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -1653,6 +1654,58 @@ def logcdf(self, value):
             tt.switch(tt.gt(y, 1), tt.log1p(-0.5 * tt.exp(-y)), tt.log(1 - 0.5 * tt.exp(-y))),
         )
 
+class AsymmetricLaplace(Continuous):
+    """
+    Assymetric Laplace Distribution
+    See also: https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution
+    """
```
Comment (ricardoV94) [inline pymc3/distributions/continuous.py:1699]: You might want to add more details in the docstrings, including the distribution parameters and respective domains, as is done for other distributions.

File: pymc3/tests/test_distributions.py [hunk 2]
```diff
@@ -206,6 +207,7 @@ def product(domains, n_samples=-1):
 
 Bool = Domain([0, 0, 1, 1], "int64")
 
+Scale = Domain([0.5,1,1.5])
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions.py]: Do you need a new Domain type? Can't the `Rplus` be used? Also the edge values should reflect the actual parameter boundaries. Is 0.5 and 1.5, the minimum and maximum values that the `scale` parameter can take?

**Review by ricardoV94 — COMMENTED** [review]
I added a few comments. I think you also need tests for the `random` method in `test_distributions_random.py`

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: > Also the edge values should reflect the actual parameter boundaries. Is 0.5 and 1.5, the minimum and maximum values that the `scale` parameter can take?

The code that we are using for asymmetric Laplace has three input parameters while the one that is taken from scipy.stats.laplace_asymmetric has only two parameters. so there are two possibilities . either we have to fix the scale to be 1 or we will have to change the referenced code. So, I thought of fixing the scale to be 1. as far as the minimum and maximum value of scale is concerned, it can be anything greater than zero. Anyway, I will change that.
we can only use `Rplus` if we modify the referenced logpdf code.
so, what should I do?

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py:1699]: okay , Thanks
I will add the docstrings.

Comment (chandan5362) [reply pymc3/distributions/continuous.py:1699]: Hey, 
I feel like changing the parameter name of `AsymmetriLaplace` to its standard value. otherwise it will create problem while writing the `math` part of docstring.
can I change the parameter name?
and also, since `laplace_asymmetric` is still not available in `scipy.stats` distribution, should I import it while writing the `plot` part of docstring ? It will save out future work.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py:1699]: I think you can skip the plot part of the docstring

Comment (ricardoV94) [reply pymc3/distributions/continuous.py:1699]: I don't know about the parameter thing. What is the problem it creates with the docstring?

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (ricardoV94) [reply pymc3/tests/test_distributions.py]: In that case I think you can create a specific Domain inside your test function (as this is unlikely to be useful for other distributions).

```python
    def test_laplace_asymmetric(self):
        self.pymc3_matches_scipy(
            AsymmetricLaplace,
            R,
            {"scale": Domain([0, 1, inf]), "symmetry": Rplus},
            laplace_asymmetric_logpdf,
        )
```

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py:1699]: basically, while writing the math fromula in latex as it is there in other classes docstrings, the standard name like `kappa` instead of `symmetry` and `b` instead of `scale` would be appropriate.

**Comment by ricardoV94** [comment]
Great work @chandan5362!

One other minor thing you should do once you fix everything is to run pre-commit to auto-format your code to be in line with the rest of the PyMC3 codebase. You have some information [here](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style), in case this is new for you. That addresses the test that is currently failing.

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: okay,  understood
Thank you

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:1699]: I don't have a strong opinion on that. I guess one argument in favor of using `b` is that the current `Laplace` distribution in Pymc3 uses `b` as well. This is kind of a running issue with PyMC, as in the `sd` vs `sigma` PR that is also open right now. Maybe someone else has a stronger opinion?

Comment (chandan5362) [reply pymc3/distributions/continuous.py:1699]: okay.
Then, I am going with `b` and `kappa` instead of `symmetry` and `scale`.
If there is any objection, please do let me know.

**Comment by chandan5362** [comment]
Hey @ricardoV94 , I have been trying to add unit test for `random` module for `laplace_asymmetric` but I ended up comparing `numpy` variable with `theano.tensor` varibale. which is giving  following error.:unamused:
`TypeError: Variables do not support boolean operations.`
Here is the code that I have been trying to add
```
  def test_laplace_asymmetric(self):
          def ref_rand(size,kappa,b):
              u = np.random.uniform(size=size)
              x = -np.log((1 - u) * (1 + kappa ** 2)) / (kappa * b) * (
                  u > ((kappa ** 2) / (1 + kappa ** 2))
              ) + kappa * np.log(
                  u * (1 + kappa ** 2) / (kappa ** 2)
              ) / b * (
                  u < ((kappa ** 2) / (1 + kappa ** 2))
              )
  
              return x
  
          pymc3_random(pm.AsymmetricLaplace,{"b":Domain([0,1,np.inf]),"kappa":Rplus}, ref_rand=ref_rand)
```
would you please help me here a little ?

File: pymc3/distributions/continuous.py [hunk 3]
```diff
@@ -1653,6 +1654,58 @@ def logcdf(self, value):
...
+
+    def random(self, point=None, size=None):
+        """
+        Draw random samples from this distribution, using the inverse CDF method.
+        :param point: not used
+        :param size: size of sample to draw
+        :return: Samples
+        """
+        if point is not None:
+            raise NotImplementedError('Random not implemented with point specified')
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: If the parameters of a random variable are fixed by the point, then there is no need to draw it in model context. This is already implemented in [pm.distributions.distribution.draw_values](https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/distribution.py#L680). So, I suspect these lines should not be there.

File: pymc3/distributions/continuous.py [hunk 4]
```diff
@@ -1653,6 +1654,58 @@ def logcdf(self, value):
...
+
+        if size is not None:
+            u = np.random.uniform(size=size)
+            x = - tt.log((1 - u) * (1 + self.symmetry ** 2)) / (self.symmetry * self.scale) * (
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: `self.symmetry` can be a random variable or just anything wrapped in [as_tensor_variable](https://theano-pymc.readthedocs.io/en/latest/library/tensor/basic.html#theano.tensor.as_tensor_variable). We certainly need to draw its values from model context.
You can do something like

```python
scale, symmetry = draw_values([self.scale, self.symmetry], point=point, size=size)
```
and then use its concrete values.

File: pymc3/distributions/continuous.py [hunk 5]
```diff
@@ -1653,6 +1654,58 @@ def logcdf(self, value):
...
+                    u > ((self.symmetry ** 2) / (1 + self.symmetry ** 2))) + self.symmetry * tt.log(
+                u * (1 + self.symmetry ** 2) / (self.symmetry ** 2)) / self.scale * (
+                        u < ((self.symmetry ** 2) / (1 + self.symmetry ** 2)))
+
+            return x
+
+        u = np.random.uniform()
+        if u > (self.symmetry ** 2) / (1 + self.symmetry ** 2):
+            x = - tt.log((1 - u) * (1 + self.symmetry ** 2)) / (self.symmetry * self.scale)
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: After we have concrete values as numpy arrays, its better to use numpy equivalents for these tensor operations. Most of the random methods in PyMC3 follow this.
Otherwise, you need a `.eval()` at the end of random method to draw values.

**Review by Sayam753 — CHANGES_REQUESTED** [review]
This looks great so far @chandan5362. Coming to the error message, you can try out the comments below and check if the test passes. 
The `pymc3_random` function checks if samples from distribution's random method and `ref_rand` come from same distribution. We need to have tests to check for shape consistencies as well. You can add a test for this very similar to how Normal distribution is tested [here](https://github.com/pymc-devs/pymc3/blob/d871b803db615f13547d6ba5a1a75387e46b5ee6/pymc3/tests/test_distributions_random.py#L313-L315).

File: pymc3/distributions/continuous.py (hunk 3 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: Thanks @Sayam753 , I kinda already emitted that and It will reflect in the upcoming PR.

**Comment by ricardoV94** [comment]
Also, you might have already done this, but if not, you should also implement the BaseTest. Here is the code for the Laplace distribution:

https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions_random.py#L373-L376

Something similar should work for your distribution. I think this takes care of testing if random works with different sizes/shapes

Edit: This was probably what @Sayam753 was saying. Sorry for the repetition :)

**Comment by chandan5362** [comment]
Thanks a lot  @Sayam753 @ricardoV94 , It was really helpful. 
Now that I know what actually was happening, I edited the `random` function and as Sayam said, values were supposed to be drawn with respect to the model, I did the same and the test is running perfectly fine.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (twiecki) [reply pymc3/distributions/continuous.py:1699]: Yeah, we're using greek letters (for better or worse) in all other distributions so should keep that consistent.

File: pymc3/tests/test_distributions.py [hunk 6]
```diff
@@ -218,7 +220,12 @@ def build_model(distfam, valuedomain, vardomains, extra_args=None):
         distfam("value", shape=valuedomain.shape, transform=None, **vals)
     return m
 
-
+def laplace_asymmetric_logpdf(value, symmetry,scale = None):
```
Comment (twiecki) [inline pymc3/tests/test_distributions.py]: Is this taken from scipy? If so I would mention that in the doc-string and also add a TODO that it should be removed once scipy adds it.

Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: > Is this taken from scipy? 
yeah @twiecki  , It has been taken from SciPy.

Comment (twiecki) [reply pymc3/tests/test_distributions.py]: OK cool, then just add those notes.

Comment (twiecki) [reply pymc3/tests/test_distributions.py]: Also not properly black formatted.

Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: hey @twiecki ,
Wish you a very very **Happy New Year**. 
please do review the PR and  let me know if anything is yet to be added or removed.

File: pymc3/distributions/continuous.py [hunk 7]
```diff
@@ -1654,6 +1655,84 @@ def logcdf(self, value):
...
+        ----------
+        value: numeric
+            Value(s) for which log-probability is calculated. If the log probabilities for multiple
+            values are desired the values must be provided in a numpy array or theano tensor
+
+        Returns
+        -------
+        TensorVariable
+        """
+        return tt.log(self.b / (self.kappa + (self.kappa ** -1))) + (
```
Comment (ricardoV94) [inline pymc3/distributions/continuous.py]: Is this function dealing properly with invalid parameters, such as a negative `b` and `kappa`? If not you might need to wrap the result in a `bound`

Comment (chandan5362) [reply pymc3/distributions/continuous.py]: I tried running this function on some  invalid parameters and It does not return any error, instead it prints the corresponding values. I guess, it can handle the invalid parameters. 
If in some sense this is wrong, please do let me know.
I am hoping that I got your question correctly.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Does it return `-inf` for invalid parameters?

Comment (chandan5362) [reply pymc3/distributions/continuous.py]: for negative `kappa`, it returns `nan`, when `kappa` is 0, it returns `-inf`. same goes for `b`

Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Then I think you should indeed wrap the results in a bound. Something like this:

```python
return bound(
    <your current return logp code here>,
    0 < b,
    0 < kappa,
)
```

This will replace the `nan` by `-inf` when using invalid negative parameters

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc3/distributions/continuous.py [hunk 8]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
         )
 
 
+class AsymmetricLaplace(Continuous):
+    r"""
+    Asymmetric-Laplace log-likelihood.
+
+    The pdf of this distribution is
+
+    ..math::
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: I do not know much about sphinx directives, but let's sync with other PyMC3 distributions.
```suggestion
    .. math::
```

File: pymc3/distributions/continuous.py [hunk 9]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+        {f(x|\\b,\kappa) =
+            \left({\frac{\\b}{\kappa + 1/\kappa}}\right)\,e^{-(x)\\b\,s\kappa ^{s}}}
+        where s = sgn(x)
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: I see another parameter `m` being used on [its wikipedia page](https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution). Are we using this `m` or any other shift parameter here?

For writing the `where` block, let's follow how its written for Gumbel distribution [here](https://github.com/pymc-devs/pymc3/blob/2824027837931a6c2f1d4ac5f210eb36b8fdf7a7/pymc3/distributions/continuous.py#L3562-L3570).

File: pymc3/distributions/continuous.py [hunk 10]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+
+    Parameters
+    ----------
+    b:
+        Scale parameter (b > 0)
+    kappa:
+        Symmetry parameter (kappa > 0)
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: I sense `b` and `kappa` are subjected to `float` values.
```suggestion
    b: float
        Scale parameter (b > 0)
    kappa: float
        Symmetry parameter (kappa > 0)
```

File: pymc3/distributions/continuous.py [hunk 11]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+
+    See also: https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: "See Also" is a separate section as per [Numpy docstring guide](https://numpydoc.readthedocs.io/en/latest/format.html#sections). Let's follow that as well.

File: pymc3/distributions/continuous.py [hunk 12]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+    """
+
+    def __init__(self, b, kappa, testval=0.0, *args, **kwargs):
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: PyMC3 base Distribution class already handles `testval` that can be passed through `kwargs`. So, I think `testval` should not be there.
While building `str` representation of a distribution/model, PyMC3 inspects all arguments passed in `__init__` method. With this way, testval may come up in the representation and can certainly be confusing to someone not familiar with test values.

File: pymc3/distributions/continuous.py [hunk 13]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+        self.b = tt.as_tensor_variable(b)
+        self.kappa = tt.as_tensor_variable(kappa)
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]:
```suggestion
        self.b = tt.as_tensor_variable(floatX(b))
        self.kappa = tt.as_tensor_variable(floatX(kappa))
```

File: pymc3/distributions/continuous.py [hunk 14]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+
+        super().__init__(*args, **kwargs, testval=testval)
+
+    def _random(self, b, kappa, size=None):
+        if size is not None:
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: I think this condition should return false when size is passed as an empty tuple. Let's go more explicit here by using `to_tuple` function.
```suggestion
        size = to_tuple(size)
        if len(size) != 0:
```

File: pymc3/tests/test_distributions.py [hunk 15]
```diff
@@ -219,6 +220,13 @@ def build_model(distfam, valuedomain, vardomains, extra_args=None):
     return m
 
 
+def laplace_asymmetric_logpdf(value, kappa, b=None):
+    kapinv = 1 / kappa
+    lPx = value * np.where(value >= 0, -kappa, kapinv)
+    lPx -= np.log(kappa + kapinv)
```
Comment (Sayam753) [inline pymc3/tests/test_distributions.py]: We need to use scale parameter `b` as well.
```suggestion
    lPx = value * b * np.where(value >= 0, -kappa, kapinv)
    lPx += np.log(b / (kappa + kapinv))
```

File: pymc3/tests/test_distributions.py [hunk 16]
```diff
@@ -958,6 +966,14 @@ def test_laplace(self):
             lambda value, mu, b: sp.laplace.logcdf(value, mu, b),
         )
 
+    def test_laplace_asymmetric(self):
+        self.pymc3_matches_scipy(
+            AsymmetricLaplace,
+            R,
+            {"b": Domain([0, 1, inf]), "kappa": Rplus},
```
Comment (Sayam753) [inline pymc3/tests/test_distributions.py]: Since both `b` and `kappa` parameters have same domains, why not use `Rplus` Domain for `b` as well?

```suggestion
            {"b": Rplus, "kappa": Rplus},
```

**Review by Sayam753 — CHANGES_REQUESTED** [review]
I addressed some changes to be made. I need some time to review the maths behind `random` method. Meanwhile you can work on the suggestions.

One more thing, you can give a mention of `AsymmetricLaplace` distribution in api source [continuous.rst](https://github.com/pymc-devs/pymc3/blob/master/docs/source/api/distributions/continuous.rst). Then we can have this distribution to show up in docs. 
There also needs to be mention of `Support`, `Mean` and `Variance` in docstring for consistency.

Also, scipy 1.6.0 has been [released](https://github.com/scipy/scipy/releases/tag/v1.6.0) a few days ago. It contains the asymmetric laplace distribution. In order to use it in test cases, we have to upgrade PyMC3 requirements. Ping @MarcoGorelli to ask if upgrading `scipy` will break anything or not.

File: pymc3/distributions/continuous.py (hunk 9 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: No, we are not using `m` here. we have considered `m` to be 0.  I am following [`this`](https://github.com/epidemics/COVIDNPIs/blob/manuscript/epimodel/pymc3_distributions/asymmetric_laplace.py) codebase as I was told to.
In case, If something is to be added, do let me know.

File: pymc3/distributions/continuous.py (hunk 12 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: Okay, Thanks 
I will have a look at it.

File: pymc3/tests/test_distributions.py (hunk 15 above)
Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: This is the code that I took from SciPy repo. Since they are using only two parameters `kappa` and `value`, I did use the same.
In case, If I am suppose to edit these, please do let me know.

File: pymc3/tests/test_distributions.py (hunk 16 above)
Comment (chandan5362) [reply pymc3/tests/test_distributions.py]: It depends if  I modify the previous suggestion. Since I was  not using `b`  in for the unit test, I thought of assigning `b` a constant value 1.
If I should  make all these changes, do let me know

File: pymc3/tests/test_distributions.py (hunk 15 above)
Comment (Sayam753) [reply pymc3/tests/test_distributions.py]: I am looking at https://scipy.github.io/devdocs/generated/scipy.stats.laplace_asymmetric.html. And it indeed uses loc, scale and kappa parameters. 
On the other hand, `b` parameter is in this distribution. So it should be tested in logpdf.

File: pymc3/tests/test_distributions.py (hunk 16 above)
Comment (Sayam753) [reply pymc3/tests/test_distributions.py]: Yes. We need to test logpdf including `b` as well since it can be parametrized in the model.

File: pymc3/distributions/continuous.py (hunk 9 above)
Comment (Sayam753) [reply pymc3/distributions/continuous.py]: That's definitely a good parameter to add. It is even there in PyMC3's [Laplace](https://docs.pymc.io/api/distributions/continuous.html#pymc3.distributions.continuous.Laplace) distribution. In order to be consistent with the given implementation, we can initialize `mu` with 0.

File: pymc3/distributions/continuous.py [hunk 17]
```diff
@@ -1654,6 +1655,93 @@ def logcdf(self, value):
...
+                u < ((kappa ** 2) / (1 + kappa ** 2))
+            )
+            return x
+
+        u = np.random.uniform()
+        if u > (kappa ** 2) / (1 + kappa ** 2):
+            x = -np.log((1 - u) * (1 + kappa ** 2)) / (kappa * b)
+        else:
+            x = kappa * np.log(u * (1 + kappa ** 2) / (kappa ** 2)) / b
+
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: The formulas are not very readable. I would suggest to go explicit, by mapping the random draws with their corresponding domains -

```suggestion
        u = np.random.uniform(size=size)
        switch = kappa ** 2 / (1 + kappa ** 2)
        non_positive_x = kappa * np.log(u * (1 / switch)) / b
        positive_x = -np.log((1 - u) * (1 + kappa ** 2)) / (kappa * b)
        draws = non_positive_x * (u <= switch) + positive_x * (u > switch)
        return draws
```
Then there will be no need of these if else conditionals.

File: pymc3/distributions/continuous.py (hunk 9 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: could you please tell me where to initialise `mu` as 0 and does it need to be a tensor variable ?

Comment (Sayam753) [reply pymc3/distributions/continuous.py]: Yes. You can use it as a default argument in `init` method. And then wrap it in `as_tensor_variable`

File: pymc3/distributions/continuous.py (hunk 12 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: Once I remove the `testval` , unit test outputs an error. I think, `testval` should be there. Please have a look at it.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: Does something like this at the end of `__init__` fix it?

```
self.mode = mu
```

Assuming you are already using mu (and have converted it to a tensor_variable)

Comment (Sayam753) [reply pymc3/distributions/continuous.py]: Interesting. I am curious to know how testval interacts `self.mode`.

Comment (ricardoV94) [reply pymc3/distributions/continuous.py]: I think it uses the mode to define the testval. If the mean or median is provided, one of those is used instead.

File: pymc3/distributions/continuous.py [hunk 18]
```diff
@@ -1661,6 +1662,103 @@ def logcdf(self, value):
...
 
+class AsymmetricLaplace(Continuous):
+    r"""
+    Asymmetric-Laplace log-likelihood.
+
+    The pdf of this distribution is
+
+    .. math::
+        {f(x|\\b,\kappa) =
+            \left({\frac{\\b}{\kappa + 1/\kappa}}\right)\,e^{-(x)\\b\,s\kappa ^{s}}}
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py]: I think this equation should include `mu` parameter as per wiki page.

File: pymc3/distributions/continuous.py [hunk 19]
```diff
@@ -1661,6 +1662,103 @@ def logcdf(self, value):
...
+
+    def __init__(self, b, kappa, mu=0, *args, **kwargs):
+        self.b = tt.as_tensor_variable(floatX(b))
+        self.kappa = tt.as_tensor_variable(floatX(kappa))
+        self.mu = mu = tt.as_tensor_variable(floatX(mu))
+
+        self.mean = self.mu - (self.kappa - 1 / self.kappa) / b
+        self.variance = (1 + self.kappa ** 4) / (self.kappa ** 2 * self.b ** 2)
+
+        super().__init__(*args, **kwargs)
```
Comment (Sayam753) [inline pymc3/distributions/continuous.py:1712]: Maybe before super method, we can check for negative values in `self.b` and `self.kappa`. Similar to how `sigma` is tested for Normal distribution [here](https://github.com/pymc-devs/pymc3/blob/982bb4d62026ade92a311895878b7f59c34bd045/pymc3/distributions/continuous.py#L491-L492).

**Review by Sayam753 — CHANGES_REQUESTED** [review]
This is turning out to be in good shape. The `random`, `logp` method and their tests need to be updated to also account for `mu` parameter.

File: pymc3/distributions/continuous.py (hunk 19 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py:1712]: okay, I will include that

File: pymc3/distributions/continuous.py (hunk 18 above)
Comment (chandan5362) [reply pymc3/distributions/continuous.py]: sure, I will add that.
Thanks though

**Comment by MarcoGorelli** [comment]
> In order to use it in test cases, we have to upgrade PyMC3 requirements. Ping @MarcoGorelli to ask if upgrading scipy will break anything or not.

My suggestion would be to first merge #4374, so that we test on the oldest supported scipy version, and then skip any tests which use a too modern scipy (see that same PR for an example) and run those tests during, say, the arviz-compat job, which uses the latest available versions of dependencies

I don't think the requirements should be updated if a feature from scipy is only needed to run the tests

**Comment by Sayam753** [comment]
Last step would be to add this distribution for docs in api source [continuous.rst](https://github.com/pymc-devs/pymc3/blob/master/docs/source/api/distributions/continuous.rst) and a mention in RELEASE-NOTES.md

**Review by Sayam753 — APPROVED** [review]

**Comment by Sayam753** [comment]
Great work @chandan5362 🎉 . Thanks @twiecki @ricardoV94 for the helpful review.

**Comment by twiecki** [comment]
What a great example of a high-quality PR and review! 💪

**Comment by chandan5362** [comment]
Thanks a lot @ricardoV94 @twiecki @Sayam753  for being helpful and supportive .
